### ISSUE TITLE
reporting missing dependencies

### ISSUE THREAD

**Issue by mfrasca** [travis-ci/travis-ci#4116]
some dependencies cannot be solved withing travis-ci, but the R code testing them does not report what is missing.

`$ Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) {   message(e);   q(status=1) }); if (!all(deps %in% installed.packages())) {  q(status = 1, save = "no") }'`

it does give an error message if something cannot be installed, but it does not report which `deps` are not in `installed.packages()`.

life (of lazy people) would be a lot easier with more explicit reporting.

https://stat.ethz.ch/R-manual/R-devel/library/base/html/sets.html

**Comment by BanzaiMan** [comment]
Paging @craigcitro @eddelbuettel @hadley

**Comment by hadley** [comment]
@BanzaiMan could we get some response to #3290? I think we're all getting frustrated with the lack of response an issue that is important to us (speeding up R-travis builds)

**Comment by mfrasca** [comment]
`Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({ deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) { message(e); q(status=1) }); if (!all(deps %in% installed.packages())) { message('missing: ', paste(setdiff(deps, installed.packages()), collapse=", ")); q(status = 1, save = "no") }'`

**Comment by mfrasca** [comment]
I don't find where is the code, I'd fork and file a pull request.

**Comment by mfrasca** [comment]
I don't find where is the code, I'd fork and file a pull request.

**Comment by craigcitro** [comment]
Travis-ci/Travis-build is the repo, happy to review (traveling atm)

On Friday, June 19, 2015, Mario Frasca <notifications@github.com> wrote:

> I don't find where is the code, I'd fork and file a pull request.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/travis-ci/travis-ci/issues/4116#issuecomment-113666126>
> .
>


-- 
-----

-cc
sent from my phone

**Comment by mfrasca** [comment]
I don't find where is the code, I'd fork it and open a pull request.

**Comment by mfrasca** [comment]
https://github.com/travis-ci/travis-build/pull/467
(sorry for multiple comments, I'm behind very feeble connection)

**Comment by craigcitro** [comment]
sorry for the slow replies (sick/traveling) -- the PR looks good to merge.

**Comment by BanzaiMan** [comment]
The PR has been merged and deployed.

**Comment by mfrasca** [comment]
I don't see the output I expected.

```bash
.
.
.
Setting up r-cran-git2r (0.10.1-1cran3ppa0precise0) ...
Setting up r-cran-devtools (1.8.0-1cran1ppa0precise0) ...
Processing triggers for libc-bin ...
ldconfig deferred processing now taking place
$ Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) {   message(e);   q(status=1) }); if (!all(deps %in% installed.packages())) {  message("missing: ", paste(setdiff(deps, installed.packages()), collapse=", "));  q(status = 1, save = "no") }'

The command "Rscript -e 'options(repos = "http://cran.rstudio.com"); tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, error = function(e) {   message(e);   q(status=1) }); if (!all(deps %in% installed.packages())) {  message("missing: ", paste(setdiff(deps, installed.packages()), collapse=", "));  q(status = 1, save = "no") }'" failed and exited with 1 during .

Your build has been stopped.
```

**Comment by machow** [comment]
I'm having the same issue as @mfrasca, even with a very simple .travis.yml.

```
language: r
sudo: required
```

is there something I missed?

**Comment by craigcitro** [comment]
adding @jimhester

**Comment by machow** [comment]
I looked more closely at the repo for hadley/devtools, tried some tweaking of the .travis.yml, and finally realized the problem is that travis-ci runs `devtools::install_deps` for all R repositories, so it's failing because there is no DESCRIPTION file in my repo.

I'm guessing there wasn't an informative error message because the `tryCatch` failed in the code below.

```
options(repos = "http://cran.rstudio.com"); 
tryCatch({   deps <- devtools::install_deps(dependencies = TRUE) }, 
               error = function(e) {   message(e);   q(status=1)}
); 
if (!all(deps %in% installed.packages())) {  
    message("missing: ", paste(setdiff(deps, installed.packages()), collapse=", "));  
    q(status = 1, save = "no") 
}
```

**Comment by jimhester** [comment]
@machow you are correct `language: r` assumes it is being run on a package repository directly. If you have some other setup you will have to use a custom solution, which you seem to have working (https://github.com/machow/2016-princetonR/blob/master/.travis.yml).

**Comment by machow** [comment]
Thanks for your response! I did, but it would be useful to avoid having to use `craigcitro/r-travis`. I ended up getting it to work with `language: r` (https://github.com/machow/2016-princetonR/blob/travis/.travis.yml). 

The problem ultimately boils down to two parts:

1. the scripts step of travis for R attempts to create the repo as a package. This can be overridden in `.travis.yml`, so it isn't much of a problem.
2. there is no way to prevent `devtools::install_deps` from running currently, so you need a DESCRIPTION file with at least one field (I think it can be arbitrary).

It seems like adding an `install_deps` option would allow for a much wider range of R repositories without needing a dummy DESCRIPTION file. It could be set to true for backwards compatibility.

The original issue is that the error message here is uninformative, though, so I'm happy to open a new issue about adding making `devtools::install_deps` optional! To be sure, I can't tell from the code why the error message wasn't informative.
